homeassistant/components/pi_hole/config_flow.py

```diff
@@ -0,0 +1,148 @@
+"""Config flow to configure the Pi-hole integration."""
+import logging
+
+from hole import Hole
+from hole.exceptions import HoleError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components.pi_hole.const import (  # pylint: disable=unused-import
+    CONF_LOCATION,
+    DEFAULT_LOCATION,
+    DEFAULT_NAME,
+    DEFAULT_SSL,
+    DEFAULT_VERIFY_SSL,
+    DOMAIN,
+)
+from homeassistant.const import (
+    CONF_API_KEY,
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PORT,
+    CONF_SSL,
+    CONF_VERIFY_SSL,
+)
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DuplicatedNameException(Exception):
+    """The name is already existed."""
+
+
+class PiHoleFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a Pi-hole config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        return await self.async_step_init(user_input)
+
+    async def async_step_import(self, user_input=None):
+        """Handle a flow initiated by import."""
+        return await self.async_step_init(user_input, is_import=True)
+
+    async def async_step_init(self, user_input, is_import=False):
+        """Handle init step of a flow."""
+        errors = {}
+
+        if user_input is not None:
+            try:
+                host = (
+                    user_input.get(CONF_HOST)
+                    if is_import
+                    else f"{user_input.get(CONF_HOST)}:{user_input.get(CONF_PORT)}"
+                )
+                name = user_input.get(CONF_NAME)
+                location = user_input.get(CONF_LOCATION)
+                tls = user_input.get(CONF_SSL)
+                verify_tls = user_input.get(CONF_VERIFY_SSL)
+                api_token = user_input.get(CONF_API_KEY)
+
+                # check unique host
+                endpoint = f"{host}/{location}"
+                existing_endpoints = [
+                    f"{entry.data.get(CONF_HOST)}/{entry.data.get(CONF_LOCATION)}"
+                    for entry in self._async_current_entries()
+                ]
+                if endpoint in existing_endpoints:
+                    return self.async_abort(reason="already_configured")
+                await self.async_set_unique_id(endpoint)
+                self._abort_if_unique_id_configured()
+
```

Comment:
`_abort_if_unique_id_configured` can `raise data_entry_flow.AbortFlow("already_configured")` so it needs to be outside of a try block